crates/attestation/src/crypto/crl.rs

```diff
@@ -0,0 +1,431 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: AGPL-3.0-only
+
+use crate::crypto::{SubjectPublicKeyInfoExt, TbsCertificateExt};
+
+use anyhow::{bail, ensure, Context, Result};
+use der::{Encode, Sequence};
+use x509::crl::CertificateList;
+use x509::name::Name;
+use x509::PkiPath;
+
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+pub struct CrlListEntry<'a> {
+    pub url: String,
+    pub crl: CertificateList<'a>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+pub struct CrlList<'a> {
+    pub crls: Vec<CrlListEntry<'a>>,
+}
+
+impl<'a> CrlList<'a> {
+    fn get_crl_by_url(&self, url: &str) -> Option<&CertificateList> {
+        for pair in &self.crls {
+            if pair.url == url {
+                return Some(&pair.crl);
+            }
+        }
+        None
+    }
+
+    fn get_crl_by_issuer(&self, issuer: &Name) -> Option<&CertificateList> {
+        for pair in &self.crls {
+            if &pair.crl.tbs_cert_list.issuer == issuer {
+                return Some(&pair.crl);
+            }
+        }
+        None
+    }
+}
+
+pub trait PkiPathCRLCheck<'a> {
+    fn check_crl(&self, pairs: &CrlList<'a>) -> Result<()>;
+}
+
+impl<'a> PkiPathCRLCheck<'a> for PkiPath<'a> {
+    fn check_crl(&self, pairs: &CrlList<'a>) -> Result<()> {
+        // We want to ensure that a valid CRL was passed, so we make sure at least one of the CRLs
+        // is valid for this `PkiPath`, otherwise, no valid CRLs were received, and that's not okay.
+        let mut found = false;
+        let mut iter = self.iter().peekable();
+        loop {
+            let cert = iter.next();
+            let next = iter.peek();
+            if cert.is_none() || next.is_none() {
+                break;
+            }
+            let cert = cert.unwrap();
+            let next = next.unwrap();
+
+            let crls = {
+                let mut crls = vec![];
+                if let Some(crl) = pairs.get_crl_by_issuer(&cert.tbs_certificate.subject) {
+                    crls.push(crl);
+                } else {
+                    let urls = cert.tbs_certificate.get_crl_urls()?;
+                    for url in urls {
+                        if let Some(crl) = pairs.get_crl_by_url(&url) {
+                            crls.push(crl);
+                        }
+                    }
+                }
+
+                crls
+            };
+
+            ensure!(!crls.is_empty(), "Certificate had no CRL URLs");
+
+            for crl in crls {
+                // Don't check CRL expiration in unit tests
+                #[cfg(not(debug_assertions))]
```

Comment:
Perhaps add a log entry (at a high level like error, or just println) when this occurs?
Just so it's obvious when this check is skipped?

Reply:
I added a print statement if `#[cfg(debug_assertions)]`.